crates/bevy_pbr/src/decal/clustered.wgsl

```diff
@@ -0,0 +1,178 @@
+fn clustered_decal_iterator_new(
+    world_position: vec3<f32>,
+    clusterable_object_index_ranges: ptr<function, ClusterableObjectIndexRanges>
+) -> ClusteredDecalIterator {
+    return ClusteredDecalIterator(
+        -1,
+        vec2(0.0),
+        // We subtract 1 because the first thing `decal_iterator_next` does is
+        // add 1.
+        i32((*clusterable_object_index_ranges).first_decal_offset) - 1,
+        i32((*clusterable_object_index_ranges).last_clusterable_object_index_offset),
+        world_position,
+    );
+}
+
+// Populates the `iterator.texture_index` and `iterator.uv` fields for the next
+// decal overlapping the current world position.
+//
+// Returns true if another decal was found or false if no more decals were found
+// for this position.
+fn clustered_decal_iterator_next(iterator: ptr<function, ClusteredDecalIterator>) -> bool {
+    if ((*iterator).decal_index_offset == (*iterator).end_offset) {
+        return false;
+    }
+
+    (*iterator).decal_index_offset += 1;
+
+    while ((*iterator).decal_index_offset < (*iterator).end_offset) {
+        let decal_index = i32(clustered_forward::get_clusterable_object_id(
+            u32((*iterator).decal_index_offset)
+        ));
+        let decal_space_vector =
+            (mesh_view_bindings::clustered_decals.decals[decal_index].local_from_world *
+            vec4((*iterator).world_position, 1.0)).xyz;
+
+        if (all(decal_space_vector >= vec3(-0.5)) && all(decal_space_vector <= vec3(0.5))) {
+            (*iterator).texture_index =
+                i32(mesh_view_bindings::clustered_decals.decals[decal_index].image_index);
+            (*iterator).uv = decal_space_vector.xy * vec2(1.0, -1.0) + vec2(0.5);
+            return true;
+        }
+
+        (*iterator).decal_index_offset += 1;
+    }
+
+    return false;
+}
+
+#endif  // CLUSTERED_DECALS_ARE_USABLE
+
+// Returns the view-space Z coordinate for the given world position.
+fn get_view_z(world_position: vec3<f32>) -> f32 {
+    return dot(vec4<f32>(
+        mesh_view_bindings::view.view_from_world[0].z,
+        mesh_view_bindings::view.view_from_world[1].z,
+        mesh_view_bindings::view.view_from_world[2].z,
+        mesh_view_bindings::view.view_from_world[3].z
+    ), vec4(world_position, 1.0));
+}
+
+// Returns true if the current view describes an orthographic projection or
+// false otherwise.
+fn view_is_orthographic() -> bool {
+    return mesh_view_bindings::view.clip_from_view[3].w == 1.0;
+}
+
+// Modifies the base color at the given position to account for decals.
+//
+// Returns the new base color with decals taken into account. If no decals
+// overlap the current world position, returns the supplied base color
+// unmodified.
+fn apply_decal_base_color(
+    world_position: vec3<f32>,
+    frag_coord: vec2<f32>,
+    initial_base_color: vec4<f32>,
+) -> vec4<f32> {
+    var base_color = initial_base_color;
+
+#ifdef CLUSTERED_DECALS_ARE_USABLE
```

Comment:
```suggestion
#ifdef CLUSTERED_DECALS_ARE_USABLE
```

Is this supposed to be called that? Do you mean to call it "CLUSTERED_DECALS_USED_IN_SCENE" or something?

Reply:
Yeah, it's supposed to be called `CLUSTERED_DECALS_ARE_USABLE`; it's `#define`'d based on whether bindless is available.

Reply:
`HAS_CLUSTERED_DECALS` or `WITH_CLUSTERED_DECALS` are more common